torch/csrc/jit/codegen/cuda/lower_thread_predicate.h

```diff
@@ -27,38 +27,50 @@ namespace cuda {
 //! If we follow a reduction parallelized on TIDx with a broadcast on TIDx we
 //! no longer need the predicate and can reset the bit accordingly
 //!
+//! In addition, if a parallel thread type is not used, it is
+//! redundant to use all threads/blocks. That isn't a problem
+//! generally although it can be inefficient, but when an aliased smem
+//! buffer is used as an output, redundant writes can be invalid (see issue
+//! #1110). PredicateInfo::redundant_types track which parallel types
+//! are redundant for each tensor and is used to let only one
+//! thread/block of a redundant type execute the expression for a
+//! tensor.
```

Comment:
Can we make smem aliasing more aggressive now then? CC @shmsong

Reply:
Yes I can try to make it more aggressive since we now explicitly guard against redundant op execution.  Meanwhile I have a question along this line, @naoyam, does this guard avoid redundant ops on `IterDomain`'s with halo as well?